src/coprocessor/cache.rs

```diff
@@ -0,0 +1,487 @@
+        let old_size = self.size;
+        self.size = res.len() + (key_size * 2) + DISTSQL_CACHE_ENTRY_ADDITION_SIZE;
+        self.version = version;
+        self.result = res;
+        self.region_id = region_id;
+        self.size - old_size
+    }
+}
+
+// RegionDistSQLCacheEntry track Region's cache items key.
+// So we can quickly get a cache entry list by given a region id.
+// That can make evict_region operation more faster than scan whole
+// cache items and get the remove item list.
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
+    // cached_items track cached DistSQL keys. So we can iterate keys
+    // for evict a region. And if we delete a cache entry we can use
+    // HashMap's remove function to fast remove a cached DistSQL key.
+    cached_items: HashMap<DistSQLCacheKey, u8>,
+}
+
+impl Default for RegionDistSQLCacheEntry {
+    fn default() -> RegionDistSQLCacheEntry {
+        RegionDistSQLCacheEntry {
+            version: 0,
+            enable: true,
+            cached_items: HashMap::default(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct DistSQLCache {
+    regions: HashMap<u64, RegionDistSQLCacheEntry>,
+    max_size: usize,
+    map: LinkedHashMap<DistSQLCacheKey, DistSQLCacheEntry>,
+    size: usize,
+}
+
+impl DistSQLCache {
+    // capacity is memory size unit is byte
+    pub fn new(capacity: usize) -> DistSQLCache {
+        DistSQLCache {
+            max_size: capacity,
+            ..Default::default()
+        }
+    }
+
+    pub fn put(&mut self, region_id: u64, k: DistSQLCacheKey, version: u64, res: Vec<u8>) {
+        if !self.is_region_cache_enabled(region_id) {
+            return;
+        }
+
+        let key_size = k.len();
+        if self.map.contains_key(&k) {
+            let mut entry = self.map.get_mut(&k).unwrap();
+            let size_diff = entry.update(region_id, version, key_size, res);
+            self.size += size_diff;
+        } else {
+            let entry = DistSQLCacheEntry::new(region_id, version, key_size, res);
+            self.size += entry.size;
+            self.map.insert(k.clone(), entry);
+            self.update_regions(region_id, k);
+        }
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    fn check_evict_key(&mut self, region_id: u64, k: &str) {
+        self.map
+            .get(k)
+            .and_then(|entry| {
+                self.regions
+                    .get(&region_id)
+                    .and_then(|rentry| Some(entry.version != rentry.version))
+            })
+            .and_then(|need_remove| {
```

Comment:
Too many `.and_then` makes me confused,  could we make it much easier to understand? @BusyJay @hicqu